README.md
# NAME

Mojolicious::Plugin::BootstrapHelpers - Type less bootstrap

<div>
    <p>
    <img src="https://img.shields.io/badge/perl-5.20+-blue.svg" alt="Requires Perl 5.20+" />
    <a href="https://travis-ci.org/Csson/p5-mojolicious-plugin-bootstraphelpers"><img src="https://api.travis-ci.org/Csson/p5-mojolicious-plugin-bootstraphelpers.svg?branch=master" alt="Travis status" /></a>
    <a href="http://cpants.cpanauthors.org/release/CSSON/Mojolicious-Plugin-BootstrapHelpers-0.0206"><img src="http://badgedepot.code301.com/badge/kwalitee/CSSON/Mojolicious-Plugin-BootstrapHelpers/0.0206" alt="Distribution kwalitee" /></a>
    <a href="http://matrix.cpantesters.org/?dist=Mojolicious-Plugin-BootstrapHelpers%200.0206"><img src="http://badgedepot.code301.com/badge/cpantesters/Mojolicious-Plugin-BootstrapHelpers/0.0206" alt="CPAN Testers result" /></a>
    <img src="https://img.shields.io/badge/coverage-87.2%-orange.svg" alt="coverage 87.2%" />
    </p>
</div>

# VERSION

Version 0.0206, released 2019-06-24.

# SYNOPSIS

    # Mojolicious
    $self->plugin('BootstrapHelpers');

    # ::Lite
    plugin 'BootstrapHelpers';

    # Meanwhile, somewhere in a template...
    %= formgroup 'Email', text_field => ['email'], large, cols => { small => [3, 9] }

    # ...that renders into
    <div class="form-group form-group-lg">
        <label class="control-label col-sm-3" for="email">Email</label>
        <div class="col-sm-9">
            <input class="form-control" id="email" name="email" type="text">
        </div>
    </div>

# STATUS

Relatively stable. This distribution will not be updated to support Bootstrap 4. There might be a separate distribution for that.

All examples are tested.

# DESCRIPTION

Mojolicious::Plugin::BootstrapHelpers is a convenience plugin that reduces some bootstrap complexity by introducing several tag helpers specifically for [Bootstrap 3](http://www.getbootstrap.com/).

The goal is not to have tag helpers for everything, but for common use cases.

All examples below (and more, see tests) is expected to work.

## How to use Bootstrap

If you don't know what Bootstrap is, see [http://www.getbootstrap.com/](http://www.getbootstrap.com/) for possible usages.

You might want to use [Mojolicious::Plugin::Bootstrap3](https://metacpan.org/pod/Mojolicious::Plugin::Bootstrap3) in your templates.

To get going quickly by using the official CDN you can use the following helpers:

    # CSS
    %= bootstrap

    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/css/bootstrap.min.css">

    # or (if you want to use the theme)
    %= bootstrap 'theme'

    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/css/bootstrap.min.css">
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/css/bootstrap-theme.min.css">

    # And the javascript
    %= bootstrap 'js'

    <script src="//maxcdn.bootstrapcdn.com/bootstrap/3.4.1/js/bootstrap.min.js"></script>

    # Or just:
    %= bootstrap 'all'

    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/css/bootstrap.min.css">
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/css/bootstrap-theme.min.css">
    <script src="//maxcdn.bootstrapcdn.com/bootstrap/3.4.1/js/bootstrap.min.js"></script>

It is also possible to automatically include jQuery (2.\*)

    %= bootstrap 'jsq'

    <script src="//code.jquery.com/jquery-2.2.4.min.js"></script>
    <script src="//maxcdn.bootstrapcdn.com/bootstrap/3.4.1/js/bootstrap.min.js"></script>

    %= bootstrap 'allq'

    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/css/bootstrap.min.css">
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/css/bootstrap-theme.min.css">
    <script src="//code.jquery.com/jquery-2.2.4.min.js"></script>
    <script src="//maxcdn.bootstrapcdn.com/bootstrap/3.4.1/js/bootstrap.min.js"></script>

## Shortcuts

There are several shortcuts for applying context and size classes that automatically expands to the correct class depending
on which tag it is applied to. For instance, if you apply the `info` shortcut to a panel, it becomes `panel-info`, but when applied to a button it becomes `btn-info`.

You can use them in two different ways, but internally they are the same. These to lines are exactly identical:

    # 'standalone shortcut'
    %= button 'Push me', primary

    # 'longform shortcut'
    %= button 'Push me', __primary => 1

For sizes, you can only use the longform (`xsmall`, `small`, `medium` and `large`). They are shortened to the Bootstrap type classes.

The following shortcuts are available:

    xsmall    default     striped       caret     right
    small     primary     bordered
    medium    success     hover
    large     info        condensed
              warning     responsive
              danger

Add two leading underscores if you don't want to use the standalone form.

See below for usage. **Important:** You can't follow a standalone shortcut with a fat comma (`=>`). The fat comma auto-quotes the string on the left, and then it breaks.

If a shortcut you try to apply isn't available in that context, it is silently not applied.

<div>
    <p>The standalone form is recommended for readability, but it does setup several helpers in your templates.
    You can turn off this style, see <a href="#init_shortcuts">init_shortcuts</a>.</p>
</div>

## Syntax convention

In the syntax sections below the following conventions are used:

    name            A specific string
    $name           Any string
    %name           One or more key-value pairs, written as:
                      key => 'value', key2 => 'value2'
                         or, if you use standalone shortcuts:
                      primary, large
    $key => [...]   Both of these are array references where the ordering of strings
    key  => [...]     are significant, for example:
                      key => [ $thing, $thing2, %hash ]
    $key => {...}   Both of these are hash references where the ordering of pairs are
    key  => {...}     are insignificant, for example:
                      key => { key2 => $value, key3 => 'othervalue' }
    (...)           Anything between parenthesis is optional. The parenthesis is not part of the
                      actual syntax
    |...|           Two pipes is a reference to another specification. For instance, button toolbars contain
                      button groups that contain buttons. Using this syntax makes the important parts clearer.
                      The pipes are not part of the actual syntax.

Ordering between two hashes that follows each other is also not significant.

**About `%has`**

The following applies to all `%has` hashes below:

- They refer to any html attributes and/or shortcuts to apply to the current element.
- When helpers are nested, all occurrencies are change to tag-specific names, such as `%panel_has`.
- This hash is always optional. It is not marked so in the definitions below in order to reduce clutter.
- Depending on context either the leading or following comma is optional together with the hash. It is usually obvious.
- Sometimes on nested helpers (such as tables in panels just below), `%has` is the only thing that can be applied to
        the other element. In this case `panel => { %panel_has }`. It follows from above that in those cases this entire
        expression is _also_ optional. Such cases are also not marked as optional in syntax definitions and are not mentioned
        in syntax description, unless they need further comment.

From this definition:

    %= table ($title,) %table_has, panel => { %panel_has }, begin
           $body
    %  end

Both of these are legal:

    # since both panel => { %panel_has } and %table_has are hashes, their ordering is not significant.
    %= table 'Heading Table', panel => { success }, condensed, id => 'the-table', begin
         <tr><td>A Table Cell</td></tr>
    %  end


    %= table begin
         <tr><td>A Table Cell</td></tr>
    %  end

## References

All other `|references|` are also helpers, so `|link|` and `|item|` needs special mention.

### |link|

`|link|` creates an `<a>` tag.

    |link|

Is exactly the same as

    $link_text, [ $url ], %link_has

**`$link_text`**

Mandatory. The text on the link.

**`$url`**

Mandatory. It sets the `href` on the link. [url\_for](https://metacpan.org/pod/Mojolicious::Controller#url_for) is used to create the link.

**`%link_has`**

Which shortcuts are available varies depending on context.

### |item|

`|item|` is used in the various submenus/dropdowns. One `|item|` creates one `<li>` tag.

    |item|

Is exactly the same as

    [ |link| ]

    # or
    $header_text

    # or
    []

So, a submenu item can be one of three things:

- 1. A link, in which case you create a `|link|` in an array reference.
- 2. A `.dropdown-header`, in which case you give it a `'string'` which then is turned into the text of the header.
- 3. A `.divider`, in which case you give it an empty array reference.

See ["Dropdowns"](#dropdowns), ["Button groups"](#button-groups) and ["Navbars"](#navbars) for examples.

# EXAMPLES

All examples below, and more, are included in html files in `/examples`. They are also available on github:

- [Badges](http://htmlpreview.github.io/?https://github.com/Csson/p5-mojolicious-plugin-bootstraphelpers/blob/master/examples/badge-1.html)
- [Include bootstrap](http://htmlpreview.github.io/?https://github.com/Csson/p5-mojolicious-plugin-bootstraphelpers/blob/master/examples/bootstrap-1.html)
- [Buttons](http://htmlpreview.github.io/?https://github.com/Csson/p5-mojolicious-plugin-bootstraphelpers/blob/master/examples/button-1.html)
- [Button groups](http://htmlpreview.github.io/?https://github.com/Csson/p5-mojolicious-plugin-bootstraphelpers/blob/master/examples/button_group-1.html)
- [Dropdowns](http://htmlpreview.github.io/?https://github.com/Csson/p5-mojolicious-plugin-bootstraphelpers/blob/master/examples/dropdown-1.html)
- [Form groups](http://htmlpreview.github.io/?https://github.com/Csson/p5-mojolicious-plugin-bootstraphelpers/blob/master/examples/formgroup-1.html)
- [Icons](http://htmlpreview.github.io/?https://github.com/Csson/p5-mojolicious-plugin-bootstraphelpers/blob/master/examples/icon-1.html)
- [Input groups](http://htmlpreview.github.io/?https://github.com/Csson/p5-mojolicious-plugin-bootstraphelpers/blob/master/examples/input_group-1.html)
- [Navs](http://htmlpreview.github.io/?https://github.com/Csson/p5-mojolicious-plugin-bootstraphelpers/blob/master/examples/nav-1.html)
- [Navbars](http://htmlpreview.github.io/?https://github.com/Csson/p5-mojolicious-plugin-bootstraphelpers/blob/master/examples/navbar-1.html)
- [Panels](http://htmlpreview.github.io/?https://github.com/Csson/p5-mojolicious-plugin-bootstraphelpers/blob/master/examples/panel-1.html)
- [Tables](http://htmlpreview.github.io/?https://github.com/Csson/p5-mojolicious-plugin-bootstraphelpers/blob/master/examples/table-1.html)
- [Toolbars](http://htmlpreview.github.io/?https://github.com/Csson/p5-mojolicious-plugin-bootstraphelpers/blob/master/examples/toolbar-1.html)

# HELPERS

## Badges

[Bootstrap documentation](http://getbootstrap.com/components/#badges)

### Syntax

    %= badge $text, %has

**`$text`**

Mandatory. If it is `undef` no output is produced.

**Available shortcuts**

`right` applies `.pull-right`.

### Examples

<div>
    A basic badge:
</div>

    <%= badge '3' %>

    <span class="badge">3</span></a>

<div>
    A right aligned badge with a data attribute:
</div>

    <%= badge '4', data => { custom => 'yes' }, right %>

    <span class="badge pull-right" data-custom="yes">4</span>

## Buttons

[Bootstrap documentation](http://getbootstrap.com/css/#buttons)

### Syntax

    %= button $button_text(, [$url]), %has

    %= submit_button $text, %has

**`$button_text`**

Mandatory. The text on the button.

**`[$url]`**

Optional array reference. It is handed off to [url\_for](https://metacpan.org/pod/Mojolicious::Controller#url_for), so with it this is
basically [link\_to](https://metacpan.org/pod/Mojolicious::Plugin::TagHelpers#link_to) with Bootstrap classes.

Not available for `submit_button`.

**Available shortcuts**

`default` `primary` `success` `info` `warning` `danger` `link` applies the various `.btn-*` classes.

`large` `small` `xsmall` applies `.btn-lg` `.btn-sm` `.btn-xs` respectively.

`active` `block` applies the `.active` and `.block` classes.

`disabled` applies the `.disabled` class if the generated element is an `<a>`. On a `<button>` it applies the `disabled="disabled"` attribute.

### Examples

<div>
    An ordinary button, with applied shortcuts:
</div>

    %= button 'The example 5' => large, warning

    <button class="btn btn-lg btn-warning" type="button">The example 5</button>

<div>
    With a url the button turns into a link:
</div>

    %= button 'The example 1' => ['http://www.example.com/'], small

    <a class="btn btn-default btn-sm" href="http://www.example.com/">The example 1</a>

<div>
    A submit button for use in forms. It overrides the build-in submit_button helper:
</div>

    %= submit_button 'Save 2', primary

    <button class="btn btn-primary" type="submit">Save 2</button>

    %= button 'Loop', active

    <button class="active btn btn-default" type="button">Loop</button>

    %= button 'Loop', block

    <button class="block btn btn-default" type="button">Loop</button>

## Button groups

### Syntax

There are two different syntaxes. One for single-button groups and one for multi-button groups. The difference is that single-button groups can't change
anything concerning the buttongroup (e.g. it can't be `justified`). If you need to do that there is nothing wrong with having a multi-button
group with just one button.

    # multi button
    <%= buttongroup %has,
                    buttons => [
                        [ |button|,
                          (items => [ |item| ])
                        ]
                    ]
    %>

    # single button
    <%= buttongroup [ |button|,
                      (items => [ |item| ])
                    ]
    %>

**`buttons => []`**

The single-button style is a shortcut for the `buttons` array reference. It takes ordinary [buttons](#buttons), with two differences: The `items` array reference, and it is unnecessary to give a button
with `items` a url.

> **`items => [...]`**
>
> Giving a button an `items` array reference consisting of one or many `|item|` creates a [dropdown](#dropdowns) like submenu. Read more under ["item"](#item).

### Examples

<div>
    A basic button group:
</div>

    <%= buttongroup
        buttons => [
            ['Button 1'],
            ['Button 2'],
            ['Button 3'],
        ]
    %>

    <div class="btn-group">
        <button class="btn btn-default" type="button">Button 1</button>
        <button class="btn btn-default" type="button">Button 2</button>
        <button class="btn btn-default" type="button">Button 3</button>
    </div>

<div>
    Nested button group. Note that the <code>small</code> shortcut is only necessary once. The same classes are automatically applied to the nested <code>.btn-group</code>:
</div>

    <%= buttongroup small,
        buttons => [
            ['Button 1'],
            ['Dropdown 1', caret, items => [
                ['Item 1', ['item1'] ],
                ['Item 2', ['item2'] ],
                [],
                ['Item 3', ['item3'] ],
            ] ],
            ['Button 2'],
            ['Button 3'],
        ],
    %>

    <div class="btn-group btn-group-sm">
        <button class="btn btn-default" type="button">Button 1</button>
        <div class="btn-group btn-group-sm">
            <button class="btn btn-default dropdown-toggle" type="button" data-toggle="dropdown">Dropdown 1 <span class="caret"></span>
            </button>
            <ul class="dropdown-menu">
                <li><a class="menuitem" href="item1" tabindex="-1">Item 1</a></li>
                <li><a class="menuitem" href="item2" tabindex="-1">Item 2</a></li>
                <li class="divider"></li>
                <li><a class="menuitem" href="item3" tabindex="-1">Item 3</a></li>
            </ul>
        </div>
        <button class="btn btn-default" type="button">Button 2</button>
        <button class="btn btn-default" type="button">Button 3</button>
    </div>

<div>
    Nested button group, with the <code>vertical</code> shortcut:
</div>

    <%= buttongroup vertical,
        buttons => [
            ['Button 1'],
            ['Dropdown 1', caret, items => [
                  ['Item 1', ['item1'] ],
                  ['Item 2', ['item2'] ],
                  [],
                  ['Item 3', ['item3'] ],
            ] ],
            ['Button 2'],
            ['Button 3'],
        ],
    %>

    <div class="btn-group-vertical">
        <button class="btn btn-default" type="button">Button 1</button>
        <div class="btn-group">
            <button class="btn btn-default dropdown-toggle" type="button" data-toggle="dropdown">Dropdown 1 <span class="caret"></span>
            </button>
            <ul class="dropdown-menu">
                <li><a class="menuitem" href="item1" tabindex="-1">Item 1</a></li>
                <li><a class="menuitem" href="item2" tabindex="-1">Item 2</a></li>
                <li class="divider"></li>
                <li><a class="menuitem" href="item3" tabindex="-1">Item 3</a></li>
            </ul>
        </div>
        <button class="btn btn-default" type="button">Button 2</button>
        <button class="btn btn-default" type="button">Button 3</button>
    </div>

<div>
    Mix links and <code>dropup</code> menus in <code>justified</code> button groups:
</div>

    <%= buttongroup justified,
        buttons => [
            ['Link 1', ['http://www.example.com/'] ],
            ['Link 2', ['http://www.example.com/'] ],
            ['Dropup 1', caret, dropup, items => [
                ['Item 1', ['item1'] ],
                ['Item 2', ['item2'] ],
                [],
                ['Item 3', ['item3'] ],
            ] ],
        ]
    %>

    <div class="btn-group btn-group-justified">
        <a class="btn btn-default" href="http://www.example.com/">Link 1</a>
        <a class="btn btn-default" href="http://www.example.com/">Link 2</a>
        <div class="btn-group dropup">
            <button class="btn btn-default dropdown-toggle" type="button" data-toggle="dropdown">Dropup 1 <span class="caret"></span>
            </button>
            <ul class="dropdown-menu">
                <li><a class="menuitem" href="item1" tabindex="-1">Item 1</a></li>
                <li><a class="menuitem" href="item2" tabindex="-1">Item 2</a></li>
                <li class="divider"></li>
                <li><a class="menuitem" href="item3" tabindex="-1">Item 3</a></li>
            </ul>
        </div>
    </div>

<div>
    Split button dropdowns uses the same syntax as any other multi-button dropdown. Set the <code>caret</code> button title to <code>undef</code>:
</div>

    <%= buttongroup
        buttons => [
            ['Link 1', ['http://www.example.com/'] ],
            [undef, caret, items => [
                ['Item 1', ['item1'] ],
                ['Item 2', ['item2'] ],
                [],
                ['Item 3', ['item3'] ],
            ] ],
        ]
    %>

    <div class="btn-group">
        <a class="btn btn-default" href="http://www.example.com/">Link 1</a>
        <div class="btn-group">
            <button class="btn btn-default dropdown-toggle" type="button" data-toggle="dropdown"><span class="caret"></span>
            </button>
            <ul class="dropdown-menu">
                <li><a class="menuitem" href="item1" tabindex="-1">Item 1</a></li>
                <li><a class="menuitem" href="item2" tabindex="-1">Item 2</a></li>
                <li class="divider"></li>
                <li><a class="menuitem" href="item3" tabindex="-1">Item 3</a></li>
            </ul>
        </div>
    </div>

<div>
    Using the simpler single-button button group dropdown syntax:
</div>

    <%= buttongroup ['Default', caret, items  => [
                        ['Item 1', ['item1'] ],
                        ['Item 2', ['item2'] ],
                        [],
                        ['Item 3', ['item3'] ],
                    ] ]
    %>

    <%= buttongroup ['Big danger', caret, large, danger, items => [
                          ['Item 1', ['item1'] ],
                          ['Item 2', ['item2'] ],
                          [],
                          ['Item 3', ['item3'] ],
                    ] ]
    %>

    <div class="btn-group">
        <button class="btn btn-default dropdown-toggle" type="button" data-toggle="dropdown">Default <span class="caret"></span>
        </button>
        <ul class="dropdown-menu">
            <li><a class="menuitem" href="item1" tabindex="-1">Item 1</a></li>
            <li><a class="menuitem" href="item2" tabindex="-1">Item 2</a></li>
            <li class="divider"></li>
            <li><a class="menuitem" href="item3" tabindex="-1">Item 3</a></li>
        </ul>
    </div>

    <div class="btn-group">
        <button class="btn btn-danger btn-lg dropdown-toggle" type="button" data-toggle="dropdown">Big danger <span class="caret"></span>
        </button>
        <ul class="dropdown-menu">
            <li><a class="menuitem" href="item1" tabindex="-1">Item 1</a></li>
            <li><a class="menuitem" href="item2" tabindex="-1">Item 2</a></li>
            <li class="divider"></li>
            <li><a class="menuitem" href="item3" tabindex="-1">Item 3</a></li>
        </ul>
    </div>

## Button toolbars

### Syntax

    <%= toolbar %toolbar_has,
                groups => [
                    { |button_group| }
                ]
    %>

**`groups => [ { |button_group| } ]`**

A mandatory array reference of [button groups](#button-groups).

### Examples

    <%= toolbar id => 'my-toolbar',
                groups => [
                    { buttons => [
                        ['Button 1'],
                        ['Button 2'],
                        ['Button 3'],
                      ],
                    },
                    { buttons => [
                        ['Button 4', primary],
                        ['Button 5'],
                        ['Button 6'],
                      ],
                    },
                ]
    %>

    <div class="btn-toolbar" id="my-toolbar">
        <div class="btn-group">
            <button class="btn btn-default" type="button">Button 1</button>
            <button class="btn btn-default" type="button">Button 2</button>
            <button class="btn btn-default" type="button">Button 3</button>
        </div>
        <div class="btn-group">
            <button class="btn btn-primary" type="button">Button 4</button>
            <button class="btn btn-default" type="button">Button 5</button>
            <button class="btn btn-default" type="button">Button 6</button>
        </div>
    </div>

## Context menus

Context menus are a slight variation on [dropdowns](#dropdowns), where the button or other thing that opens the menu isn't part of the menu, such as menus that
opens on right click. The context menu is created without the <div class="dropdown"></div> wrapper, so for it to show up at the right place its
position must be set using either CSS or Javascript.

### Syntax

    <%= context_menu %has, items  => [ |item| ] %>

**`items`**

Mandatory array reference consisting of one or many `|item|`. Read more under ["item"](#item).

### Examples

        <%= context_menu id => 'my-context-menu', items => [
                ['Item 1', ['item1'] ],
                ['Item 2', ['item2'] ],
                [],
                ['Item 3', ['item3'] ]
             ] %>
    </div>

    <ul class="dropdown-menu" id="my-context-menu">
        <li><a class="menuitem" href="item1" tabindex="-1">Item 1</a></li>
        <li><a class="menuitem" href="item2" tabindex="-1">Item 2</a></li>
        <li class="divider"></li>
        <li><a class="menuitem" href="item3" tabindex="-1">Item 3</a></li>
    </ul>

## Dropdowns

### Syntax

    <%= dropdown  %has,
                  [ |button|, items  => [ |item| ]
                  ]

**`[ |button| ]`**

Mandatory array reference. It takes an ordinary [button](#buttons), with two differences: The `items` array reference, and it is unnecessary to give a button
with `items` a url.

> **`items`**
>
> Mandatory array reference consisting of one or many `|item|`. Read more under ["item"](#item).

**Available shortcuts**

`caret` adds a `<span class="caret"></span<>` element on the button.

### Examples

<div>
    By default, <code>tabindex</code> is set to <code>-1</code>:
</div>

    <div class="text-right">
        <%= dropdown
             ['Dropdown 1', id => 'a_custom_id', right, items => [
                ['Item 1', ['item1'] ],
                ['Item 2', ['item2'] ],
                [],
                ['Item 3', ['item3'] ]
             ] ] %>
    </div>

    <div class="text-right">
        <div class="dropdown">
            <button class="btn btn-default dropdown-toggle" type="button" id="a_custom_id" data-toggle="dropdown">Dropdown 1</button>
            <ul class="dropdown-menu dropdown-menu-right">
                <li><a class="menuitem" href="item1" tabindex="-1">Item 1</a></li>
                <li><a class="menuitem" href="item2" tabindex="-1">Item 2</a></li>
                <li class="divider"></li>
                <li><a class="menuitem" href="item3" tabindex="-1">Item 3</a></li>
            </ul>
        </div>
    </div>

<div>
    ...but it can be overridden:
</div>

    <%= dropdown
         ['Dropdown 2', caret, large, primary, items => [
            ['Item 1', ['item1'], data => { attr => 2 } ],
            ['Item 2', ['item2'], disabled, data => { attr => 4 } ],
            [],
            ['Item 3', ['item3'], data => { attr => 7 } ],
            [],
            ['Item 4', ['item4'], tabindex => 4 ],
            'This is a header',
            ['Item 5', ['item5'] ],
         ] ] %>

    <div class="dropdown">

Changes
Revision history for Mojolicious-Plugin-BootstrapHelpers

{{$NEXT}}
 [API Changes]

 [Bug Fixes]

 [Documentation]

 [Enhancements]

 [New Features]

 [Other Changes]

 [Requirements]

 [Stats & Dependency Changes]

 [Testing]

0.0206 2019-06-24T20:14:11Z UTC
 [Enhancements]
 - Updated Bootstrap version to 3.4.1 (when included via the cdn helpers).

 [Stats & Dependency Changes]
 - 4 files changed, 42 insertions(+), 42 deletions(-)

0.0205 2018-12-31T19:02:24Z UTC
 [Enhancements]
 - Updated Bootstrap version to 3.4.0 (when included via the cdn helpers).

 [Stats & Dependency Changes]
 - 4 files changed, 35 insertions(+), 33 deletions(-)
 - (dev req) ~ Dist::Zilla 6.008 --> 6.012
 - (dev req) ~ Dist::Zilla::Plugin::BumpVersionAfterRelease::Transitional
   0.007 --> 0.008
 - (dev req) ~ Dist::Zilla::Plugin::CopyFilesFromBuild 0.163040 -->
   0.170880
 - (dev req) ~ Dist::Zilla::Plugin::Git 2.041 --> 2.045
 - (dev req) ~ Dist::Zilla::Plugin::Git::Contributors 0.029 --> 0.034
 - (dev req) ~ Dist::Zilla::Plugin::MetaProvides::Package 2.003001 -->
   2.004003
 - (dev req) ~ Dist::Zilla::Plugin::PromptIfStale (any) --> 0.055
 - (dev req) ~ Dist::Zilla::Plugin::Test::Compile 2.056 --> 2.057
 - (dev req) ~ Dist::Zilla::Plugin::Test::ReportPrereqs 0.021 --> 0.027
 - (dev sug) ~ Dist::Iller::Config::Author::CSSON 0.0313 --> 0.0322

0.0204 2016-12-31T16:43:08Z UTC
 [Documentation]
 - Updated documentation regarding which versions of Bootstrap and jQuery
   are included when using those helpers.

 [Stats & Dependency Changes]
 - 2 files changed, 15 insertions(+), 15 deletions(-)
 - (dev req) ~ Dist::Zilla 5.000 --> 6.008
 - (dev req) ~ Dist::Zilla::Plugin::CopyFilesFromBuild 0.151680 -->
   0.163040
 - (dev req) ~ Dist::Zilla::Plugin::Git 2.036 --> 2.041
 - (dev req) ~ Dist::Zilla::Plugin::Git::CheckFor::CorrectBranch 0.013 -->
   0.014
 - (dev req) ~ Dist::Zilla::Plugin::Git::Contributors 0.017 --> 0.029
 - (dev req) ~ Dist::Zilla::Plugin::ReadmeAnyFromPod 0.150250 --> 0.163250
 - (dev req) ~ Dist::Zilla::Plugin::Test::Compile 2.054 --> 2.056
 - (dev sug) ~ Dist::Iller::Config::Author::CSSON 0.0312 --> 0.0313

0.0203 2016-09-20T21:16:40Z UTC
 [New Features]
 - Added the 'context_menu' helper, a dropdown menu without button.

 [Stats & Dependency Changes]
 - 4 files changed, 109 insertions(+), 12 deletions(-)
 - (dev req) ~ Dist::Zilla::Plugin::ChangeStats::Dependencies::Git 0.0100
   --> 0.0200

0.0202 2016-08-16T20:42:05Z UTC
 [Enhancements]
 - Updated Bootstrap version to 3.3.7 (when included via the cdn helpers).
 - Updated jQuery version to 2.2.4 (when included via the cdn helpers).

 [Other]
 - No changes necessary for Mojolicious 7 compatibility.

 [Stats & Dependency Changes]
 - 4 files changed, 38 insertions(+), 38 deletions(-)
 - (dev req) + Dist::Zilla::Plugin::MetaConfig (any)
 - (dev req) + Dist::Zilla::Plugin::PromptIfStale (any)
 - (dev sug) ~ Dist::Iller::Config::Author::CSSON 0.0309 --> 0.0312

0.0201 2016-03-23T22:59:42Z UTC
 [Bug Fixes]
 - Fixed an issue where an item in a 'nav' that wasn't a dropdown couldn't
   have attributes.

 [Documentation]
 - Removed duplicated pod sections.

 [Enhancements]
 - Added 'container' argument to navbar.
 - Updated Bootstrap version to 3.3.6 (when included via the cdn helpers).
 - Updated jQuery version to 2.2.2 (when included via the cdn helpers).

 [Other Changes]
 - No longer uses package block syntax.

 [Requirements]
 - Requires latest (1.44) List::Util instead of List::MoreUtils, since it
   now has 'uniq'.

 [Stats & Dependency Changes]
 - 9 files changed, 961 insertions(+), 821 deletions(-)
 - (run req) + List::Util 1.44
 - (run req) - List::MoreUtils
 - (test req) + IO::Handle (any)
 - (test req) + IPC::Open3 (any)
 - (dev req) + Badge::Depot 0.0103
 - (dev req) + Badge::Depot::Plugin::Coverage (any)
 - (dev req) + Badge::Depot::Plugin::Cpantesters (any)
 - (dev req) + Badge::Depot::Plugin::Kwalitee (any)
 - (dev req) + Badge::Depot::Plugin::Perl (any)
 - (dev req) + Badge::Depot::Plugin::Travis (any)
 - (dev req) + Dist::Zilla 5.000
 - (dev req) + Dist::Zilla::Plugin::Authority 1.009
 - (dev req) + Dist::Zilla::Plugin::BumpVersionAfterRelease::Transitional
   0.007
 - (dev req) + Dist::Zilla::Plugin::ChangeStats::Dependencies::Git 0.0100
 - (dev req) + Dist::Zilla::Plugin::Clean 0.07
 - (dev req) + Dist::Zilla::Plugin::Git 2.036
 - (dev req) + Dist::Zilla::Plugin::Git::CheckFor::CorrectBranch 0.013
 - (dev req) + Dist::Zilla::Plugin::Git::Commit (any)
 - (dev req) + Dist::Zilla::Plugin::InstallRelease 0.008
 - (dev req) + Dist::Zilla::Plugin::NextRelease::Grouped 0.0200
 - (dev req) + Dist::Zilla::Plugin::RewriteVersion::Transitional 0.007
 - (dev req) + Dist::Zilla::Plugin::Test::Compile 2.054
 - (dev req) + Dist::Zilla::Plugin::Test::EOF 0.0501
 - (dev req) + Dist::Zilla::Plugin::Test::EOL 0.18
 - (dev req) + Dist::Zilla::Plugin::Test::Kwalitee::Extra v0.2.1
 - (dev req) + Dist::Zilla::Plugin::Test::NoTabs 0.15
 - (dev req) + Dist::Zilla::Plugin::Test::Version 1.05
 - (dev req) + Pod::Weaver::Section::GenerateSection 1.01
 - (dev req) + Pod::Weaver::Section::Legal::Supplemented (any)
 - (dev req) + Test::EOF (any)
 - (dev req) + Test::EOL (any)
 - (dev req) + Test::More 0.88
 - (dev req) + Test::NoTabs 1.4
 - (dev req) + Test::Version 1
 - (dev req) ~ Dist::Zilla::Plugin::CheckChangesHasContent (any) --> 0.008
 - (dev req) ~ Dist::Zilla::Plugin::CopyFilesFromBuild (any) --> 0.151680
 - (dev req) ~ Dist::Zilla::Plugin::Git::Contributors (any) --> 0.017
 - (dev req) ~ Dist::Zilla::Plugin::GithubMeta (any) --> 0.54
 - (dev req) ~ Dist::Zilla::Plugin::MetaProvides::Package (any) -->
   2.003001
 - (dev req) ~ Dist::Zilla::Plugin::PodWeaver (any) --> 4.006
 - (dev req) ~ Dist::Zilla::Plugin::ReadmeAnyFromPod (any) --> 0.150250
 - (dev req) ~ Dist::Zilla::Plugin::RunExtraTests (any) --> 0.028
 - (dev req) ~ Dist::Zilla::Plugin::Stenciller::HtmlExamples (any) -->
   0.0201

Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.012.
use strict;
use warnings;

use 5.020000;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Type less bootstrap",
  "AUTHOR" => "Erik Carlsson <info\@code301.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Mojolicious-Plugin-BootstrapHelpers",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.020000",
  "NAME" => "Mojolicious::Plugin::BootstrapHelpers",
  "PREREQ_PM" => {
    "List::Util" => "1.44",
    "Mojolicious" => "5.0",
    "Scalar::Util" => "1.29",
    "String::Random" => 0,
    "String::Trim" => "0.004",
    "experimental" => "0.008"
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::Mojo::Trim" => "0.03",
    "Test::More" => "0.96"
  },
  "VERSION" => "0.0207",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "List::Util" => "1.44",
  "Mojolicious" => "5.0",
  "Scalar::Util" => "1.29",
  "String::Random" => 0,
  "String::Trim" => "0.004",
  "Test::Mojo::Trim" => "0.03",
  "Test::More" => "0.96",
  "experimental" => "0.008"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
